pkg/controller/controller.go

```diff
@@ -475,9 +476,49 @@ func (c *M3DBController) handleClusterUpdate(cluster *myspec.M3DBCluster) error
 		}
 	}
 
+	// At this point we have the desired number of statefulsets, and every pod
+	// across those sets is bootstrapped. However some may be bootstrapped because
+	// they own no shards. Check to see that all pods are in the placement.
+	clusterPodsSelector := klabels.SelectorFromSet(labels.BaseLabels(cluster))
+	pods, err := c.podLister.Pods(cluster.Namespace).List(clusterPodsSelector)
+	if err != nil {
+		return fmt.Errorf("could not list pods to check update revision: %v", err)
+	}
+
+	// For all statefulsets, ensure their observered generation is up to date.
+	// This means that the k8s statefulset controller has updated Status (and
+	// therefore ready replicas + updated replicas). If observed generation !=
+	// generation, it means Status will contain stale info.
+	for _, sts := range childrenSets {
+		if sts.Generation != sts.Status.ObservedGeneration {
+			c.logger.Warn("stateful set not up to date",
+				zap.String("namespace", sts.Namespace),
+				zap.String("name", sts.Name),
+				zap.Int32("readyReplicas", sts.Status.ReadyReplicas),
+				zap.Int32("updatedReplicas", sts.Status.UpdatedReplicas),
+				zap.String("currentRevision", sts.Status.CurrentRevision),
+				zap.String("updateRevision", sts.Status.UpdateRevision),
+				zap.Int64("generation", sts.Generation),
+				zap.Int64("observed", sts.Status.ObservedGeneration),
+			)
+			return fmt.Errorf("set %s generation is not up to date (current: %d, observed: %d)", sts.Name, sts.Generation, sts.Status.ObservedGeneration)
+		}
+	}
+
 	// If any of the statefulsets aren't ready, wait until they are as we'll get
 	// another event (ready == bootstrapped)
 	for _, sts := range childrenSets {
+		c.logger.Info("processing set",
```

Comment:
Still need to decide whether these logs are too noisy. If we had them, however, debugging this issue would have been much easier.

Reply:
Can you simply make these debug logs and then make it easy to turn on debug logging? (might already be easy by just changing the config file?)